packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/RealmWebSocketTransport.kt

```diff
@@ -0,0 +1,123 @@
+package io.realm.kotlin.mongodb.internal
+
+import io.realm.kotlin.internal.interop.RealmWebsocketHandlerCallbackPointer
+import io.realm.kotlin.internal.interop.sync.CancellableTimer
+import io.realm.kotlin.internal.interop.sync.WebSocketClient
+import io.realm.kotlin.internal.interop.sync.WebSocketObserver
+import io.realm.kotlin.internal.interop.sync.WebSocketTransport
+import io.realm.kotlin.internal.interop.sync.WebsocketEngine
+import io.realm.kotlin.internal.util.CoroutineDispatcherFactory
+import io.realm.kotlin.internal.util.DispatcherHolder
+import kotlinx.atomicfu.AtomicRef
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+public class RealmWebSocketTransport(
+    public val timeoutMs: Long
+) : WebSocketTransport {
+    // we need a single thread dispatcher to act like an event loop
+    private val dispatcherHolder: DispatcherHolder = CoroutineDispatcherFactory.managed("RealmWebsocketTransport").create()
+    public val scope: CoroutineScope = CoroutineScope(dispatcherHolder.dispatcher)
+
+    public val engine: WebsocketEngine by lazy { websocketEngine(timeoutMs) }
+
+    private val websocketClients = mutableListOf<WebSocketClient>()
+
+    override fun post(handlerCallback: RealmWebsocketHandlerCallbackPointer) {
+        scope.launch {
+            (this as Job).invokeOnCompletion { completionHandler: Throwable? ->
```

Comment:
Because the scheduled coroutine could have been cancelled, depending on this we run different callback from Core's perspective. hence the usage of `invokeOnCompletion` which reports if the execution was normal or cancelled (subject to unhandled error)